content/blog/2023/2023-10-25-spring-java-config.md

```diff
@@ -0,0 +1,1176 @@
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+public class MyController {
+    @GetMapping("/hello")
+    public String hello() {
+        return "Hello, Spring Boot!";
+    }
+}
+```
+In this example, a simple REST endpoint is created without explicit configuration. Spring Boot's auto-configuration handles the setup, allowing developers to focus on the business logic.
+
+Spring Boot's auto-configuration simplifies `@RestController` development. It handles `HTTP` message conversion, request mapping, and exception handling. Developers can focus on business logic while Spring Boot manages content negotiation and routing, streamlining RESTful API creation with sensible defaults and minimizing manual configuration.
+
+In the context of a `@RestController`, Spring Boot offers several auto-configurations that simplify the setup and handling of RESTful endpoints. These auto-configurations are part of Spring Boot's opinionated defaults, allowing developers to quickly create REST APIs without having to configure everything manually.
+
+Here's what Spring Boot provides in terms of auto-configuration for a `@RestController`:
+
+1. HTTP Message Conversion:
+
+Spring Boot automatically configures message converters to handle the conversion between Java objects and `JSON` (and other media types). When your controller returns an object, Spring Boot automatically serializes it to `JSON` using libraries like Jackson.
+
+Example:
+
+```java
+@RestController
+public class MyController {
+
+    @GetMapping("/hello")
+    public ResponseEntity<String> hello() {
+        // Automatically converted to JSON
+        return ResponseEntity.ok("Hello, Spring Boot!");
+    }
+}
+```
+
+2. Request Mapping and Dispatching:
+
+Spring Boot sets up default request mappings and dispatcher servlet configurations. It automatically maps incoming `HTTP` requests to the appropriate controller methods based on the `@RequestMapping` or related annotations.
+
+Example:
+
+```java
+@RestController
+@RequestMapping("/api")
+public class MyController {
+
+    @GetMapping("/hello")
+    public ResponseEntity<String> hello() {
+        // Maps to /api/hello
+        return ResponseEntity.ok("Hello, Spring Boot!");
+    }
+}
+```
+
+3. Exception Handling:
+
+Spring Boot provides default exception handling, mapping exceptions to appropriate `HTTP` status codes and error responses. For example, if a method throws an exception, Spring Boot automatically generates a `500` `Internal Server Error` response.
+
+Example:
+
+```java
+@RestController
+public class MyController {
+    @GetMapping("/hello")
+    public ResponseEntity<String> hello() {
+        // Some logic that might throw an exception
+        throw new RuntimeException("Something went wrong!");
+    }
+}
+```
+
+In this case, Spring Boot would automatically handle the exception and send a `500` `Internal Server Error` response to the client.
+
+4. Content Negotiation
```

Comment:
This doesn't really have anything to do with "Java Config", does it?

Reply:
Removed it to avoid confusion.